### DISCUSSION TITLE
:pray: Better howto for `EXTRA_HTTP_HEADERS` and `HTTP` auth

### DISCUSSION THREAD

**Discussion by adriens** [duckdb/duckdb#14165] [Ideas]
# :grey_question: About

[Secret manager](https://duckdb.org/docs/configuration/secrets_manager.html) made it possible to [Query an Authenticated API Endpoint](https://duckdbsnippets.com/snippets/184/query-an-authenticated-api-endpoint) : 

![image](https://github.com/user-attachments/assets/725e7e96-4e0f-4b68-88f8-81a6f0ec8354)

So today, I'm trying to call one of my APIs but I'm stuck on `401` while the `http` call runs perfectly.

**:point_right: The aim of this discussion is to provide some more documentation on how to achieve this** 

# :scroll: Steps to reproduce

Assuming the following `httpie` script is ok : 

```shell
http GET https://api.opt.nc/temps-attente-agences/agences x-apikey:uskjhkjhqkjsfhkqjhfkjqhfkjdqhfkqgfiohjsd

```
Which SQL statement should be run to avoid a `401` to : 

- Create the `SECRET`
- `SELECT` an `API` endpoint

# :bookmark_tabs: Related resources

- https://github.com/duckdb/duckdb/discussions/6782
- [Tweet](https://x.com/siddiqueESL/status/1839462968968311021)

**Comment by adriens** [comment]
# :ok: Current code snippet output

![image](https://github.com/user-attachments/assets/7cbaef39-7173-47c6-b795-a8cff8932720)

**Comment by adriens** [comment]
## :sos: Current custom API call usecase

![Screenshot from 2024-09-30 08-59-01](https://github.com/user-attachments/assets/cfb566b5-d6ff-4603-8d3e-29175321d028)

### :package:  `json` output sample

```json
[
  {
    "codeESirius": "site6",
    "codePostal": "98808",
    "codePostalRefloc": "98800",
    "commune": "NOUMEA",
    "coordonneeX": 446400.6599999964,
    "coordonneeXPrecise": 446400,
    "coordonneeY": 218117.88000000082,
    "coordonneeYPrecise": 218117,
    "designation": "Centre de Traitement Postal",
    "idAgence": 1151,
    "idRefloc": "05ADNC0118025400150000",
    "lieuDitOuTribu": "",
    "localite": "NOUMEA CEDEX",
    "localiteRefloc": "NOUMEA",
    "position": {
      "lat": -22.238939402080792,
      "lon": 166.45011686307052
    },
    "realMaxWaitingTimeMs": 0,
    "type": "CTC"
  },
  {
    "codeESirius": "site12",
    "codePostal": "98841",
    "codePostalRefloc": "98800",
    "commune": "NOUMEA",
    "coordonneeX": 445934.3800000027,
    "coordonneeXPrecise": 445927,
    "coordonneeY": 215750.6400000006,
    "coordonneeYPrecise": 215749,
    "designation": "Agence de NOUMEA VALLEE DU TIR",
    "idAgence": 1152,
    "idRefloc": "05ADNC0118006500030000",
    "lieuDitOuTribu": "",
    "localite": "NOUMEA CEDEX",
    "localiteRefloc": "NOUMEA",
    "position": {
      "lat": -22.260329590749315,
      "lon": 166.4456591312507
    },
    "realMaxWaitingTimeMs": 0,
    "type": "AGENCE"
  },
  {
    "codePostal": "98889",
    "codePostalRefloc": "98889",
    "commune": "PAITA",
    "coordonneeX": 439905.5300000012,
    "coordonneeXPrecise": 0,
    "coordonneeY": 226468.01400000043,
    "coordonneeYPrecise": 0,
    "designation": "LAVERIE DES SENTEURS",
    "idAgence": 10851,
    "idRefloc": "TMP39985",
    "lieuDitOuTribu": "LOTISSEMENT SAVANNAH",
    "localite": "PAITA",
    "localiteRefloc": "PAITA",
    "position": {
      "lat": -22.163686576300545,
      "lon": 166.38690963624717
    },
    "realMaxWaitingTimeMs": 0,
    "type": "POINT_RELAIS"
  },
  {
    "codeESirius": "",
    "codePostal": "98841",
    "codePostalRefloc": "98800",
    "commune": "NOUMEA",
    "coordonneeX": 445672.8500000015,
    "coordonneeXPrecise": 0,
    "coordonneeY": 214091.58000000007,
    "coordonneeYPrecise": 0,
    "designation": "AGENCE COMPTABLE - CAVEAU",
    "idAgence": 14051,
    "idRefloc": "05ADNC0118049900070000",
    "lieuDitOuTribu": "",
    "localite": "NOUMEA CEDEX",
    "localiteRefloc": "NOUMEA",
    "position": {
      "lat": -22.275318886068586,
      "lon": 166.4431673905479
    },
    "realMaxWaitingTimeMs": 0,
    "type": "AGENCE"
  }
]
```

**Comment by Tishj** [comment]
This sounds like a documentation issue (/request?)
Perhaps @szarnyasg can move it there (possibly upgrade it to an issue ?)

**Reply by adriens** [reply]
Of course 
Maybe an issue would be more suitable

**Comment by Tishj** [comment]
To answer your question
Your key should likely be `X-API-Key`, and your value `uskjhkjhqkjsfhkqjhfkjqhfkjdqhfkqgfiohjsd`

Reading this: <https://swagger.io/docs/specification/v3_0/authentication/api-keys/>

------------------------------------------

Also <https://docs.stripe.com/payments/forwarding-token-vault#create-forwarding-request> helps to see that they are separated into `name` and `value` for the created request headers

**Reply by Tishj** [reply]
> Your key should likely be `X-API-Key`, and your value `uskjhkjhqkjsfhkqjhfkjqhfkjdqhfkqgfiohjsd`

@adriens did this work or not?
```sql
    EXTRA_HTTP_HEADERS MAP {
        'X-API-Key': 'uskjhkjhqkjsfhkqjhfkjqhfkjdqhfkqgfiohjsd'
    }
```

**Reply by adriens** [reply]
![Screenshot from 2024-10-01 07-22-05](https://github.com/user-attachments/assets/51e07d31-f211-4d0e-b4d7-38d0c5d6b4af)

**Reply by Tishj** [reply]
That's not what I suggested

**Reply by Tishj** [reply]
It looks like the API you're using is rejecting the HEAD request, can you verify if the HEAD request is also breaking when you use `httpie` ?

**Reply by adriens** [reply]
> It looks like the API you're using is rejecting the HEAD request,

Indeed, yes

> can you verify if the HEAD request is also breaking when you use httpie ?

Putting paramaters like [`Key:value` in httpie](https://httpie.io/docs/cli/http-headers) should already load into headers or so I thought and it's ok from python, see below a working source code ; 

```python
apikey = load_apikey()
headers = {
    "x-apikey": apikey,
    "Content-Type": "application/json"
}

API_TEMPS_ATTENTE_BASE_URL = "https://api.opt.nc/temps-attente-agences/"

# requête pour récuperer les commune
response_communes = requests.get(API_TEMPS_ATTENTE_BASE_URL+"communes", headers=headers)
```

Don't really know how to call with `httpie` according to your requests : can you give me some example ?

**Reply by Tishj** [reply]
I'm not responding to this further, this type of remote debugging is not very fruitful
Your code has `http GET ` I imagine there is a way to do `http HEAD ` instead.

It doesn't look like a duckdb issue as the HEAD request is being sent out with the right headers, the API you're using is likely just rejecting HEAD requests to that endpoint

The docs have a tip that might be applicable here:
> If you get an IO Error (Connection error for HTTP HEAD), configure the endpoint explicitly via ENDPOINT 's3.⟨your-region⟩.amazonaws.com'.

from <https://duckdb.org/docs/extensions/httpfs/s3api.html>

**Reply by adriens** [reply]
>  the API you're using is likely just rejecting HEAD requests to that endpoint

It's hosted behind apigee, I'm looking further and will let know.

**Comment by adriens** [comment]
Hi @Tishj we're working on upgrading our API Gateway configuration.
I will let you know once it's done and make some tweet about it.

**Comment by adriens** [comment]
:partying_face: 
![image](https://github.com/user-attachments/assets/975982ce-dd5a-4bb4-b316-ebee5d74fcef)

**Comment by adriens** [comment]
[Tweet](https://x.com/rastadidi/status/1843208376005832824) @Tishj :pray: 

![image](https://github.com/user-attachments/assets/53848e29-3a88-4f7b-a99b-fed9152f59bb)
